lib/rules/no-useless-assignment.js

```diff
@@ -0,0 +1,564 @@
+         */
+        function verify(target) {
+
+            /**
+             * Checks whether the given identifier is used in the segment.
+             * @param {CodePathSegment} segment The code path segment.
+             * @param {Identifier} identifier The identifier to check.
+             * @returns {boolean} `true` if the identifier is used in the segment.
+             */
+            function isIdentifierUsedInSegment(segment, identifier) {
+                const segmentInfo = target.segments[segment.id];
+
+                return (
+                    segmentInfo.first &&
+                    segmentInfo.last &&
+                    segmentInfo.first.range[0] <= identifier.range[0] &&
+                    identifier.range[1] <= segmentInfo.last.range[1]
+                );
+            }
+
+            /**
+             * Verifies whether the given assignment info is an unused assignment.
+             * Report if it is an unused assignment.
+             * @param {AssignmentInfo} targetAssignment The assignment info to verify.
+             * @param {AssignmentInfo[]} allAssignments The list of all assignment info for variables.
+             * @returns {void}
+             */
+            function verifyAssignmentInfoIsUnused(targetAssignment, allAssignments) {
+
+                /**
+                 * @typedef {Object} SubsequentSegmentData
+                 * @property {CodePathSegment} segment The code path segment
+                 * @property {AssignmentInfo} [assignment] The first occurrence of the assignment within the segment.
+                 * There is no need to check if the variable is used after this assignment,
+                 * as the value it was assigned will be used.
+                 */
+
+                /**
+                 * Information used in `getSubsequentSegments()`.
+                 * To avoid unnecessary iterations, cache information that has already been iterated over,
+                 * and if additional iterations are needed, start iterating from the retained position.
+                 */
+                const subsequentSegmentData = {
+
+                    /**
+                     * Cache of subsequent segment information list that have already been iterated.
+                     * @type {SubsequentSegmentData[]}
+                     */
+                    results: [],
+
+                    /**
+                     * Subsequent segments that have already been iterated on. Used to avoid infinite loops.
+                     * @type {Set<CodePathSegment>}
+                     */
+                    subsequentSegments: new Set(),
+
+                    /**
+                     * Unexplored code path segment.
+                     * If additional iterations are needed, consume this information and iterate.
+                     * @type {CodePathSegment[]}
+                     */
+                    queueSegments: [...targetAssignment.segment.nextSegments]
```

Comment:
`nextSegments` of a `segment` never include unreachable segments, which is great as they should be skipped when we're analyzing a reachable `segment`.

However, when the `segment` itself is unreachable, analysis based on `nextSegments` would be wrong. For example:

```js
/* eslint no-useless-assignment: "error" */

() => {
    return;

    const x = 1; // The rule reports here, but this is unreachable. If this were reachable, the assignment would be valid.
    if (foo) {
        bar(x);
    }
}

```

I think the rule should just ignore assignments in unreachable segments.